pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/typedefinition/JavaTypeDefinition.java

```diff
@@ -106,254 +84,80 @@ public static int getGenericTypeIndex(TypeVariable<?>[] typeParameters, final St
-        return false;
-    }
+    public abstract boolean isClassOrInterface();
 
-    public boolean isPrimitive() {
-        return clazz.isPrimitive();
-    }
+    public abstract boolean isNullType();
 
-    public boolean equivalent(JavaTypeDefinition def) {
-        // TODO: JavaTypeDefinition generic equality
-        return clazz.equals(def.clazz) && getTypeParameterCount() == def.getTypeParameterCount();
-    }
+    public abstract boolean isPrimitive();
 
-    public boolean hasSameErasureAs(JavaTypeDefinition def) {
-        return clazz == def.clazz;
-    }
+    public abstract boolean hasSameErasureAs(JavaTypeDefinition def);
 
-    public int getTypeParameterCount() {
-        return typeParameterCount;
-    }
+    public abstract int getTypeParameterCount();
 
-    public boolean isArrayType() {
-        return clazz.isArray();
-    }
+    public abstract boolean isArrayType();
 
     @Override
-    public String toString() {
-        return new StringBuilder("JavaTypeDefinition [clazz=").append(clazz)
-                .append(", genericArgs=").append(genericArgs)
-                .append(", isGeneric=").append(isGeneric)
-                .append(']').toString();
-
-    }
+    public abstract String toString();
 
     @Override
-    public boolean equals(Object obj) {
-        if (obj == null || !(obj instanceof JavaTypeDefinition)) {
-            return false;
-        }
-
-        // raw vs raw
-        // we assume that this covers raw types, because they are cached
-        if (this == obj) {
-            return true;
-        }
-
-        JavaTypeDefinition otherTypeDef = (JavaTypeDefinition) obj;
-
-        if (clazz != otherTypeDef.clazz) {
-            return false;
-        }
+    public abstract boolean equals(Object obj);
 
+    @Override
+    public abstract int hashCode();
 
-        // This should cover
-        // raw vs proper
-        // proper vs raw
-        // proper vs proper
+    public abstract Set<JavaTypeDefinition> getSuperTypeSet();
 
-        // Note: we have to force raw types to compute their generic args, class Stuff<? extends List<Stuff>>
-        // Stuff a;
-        // Stuff<? extends List<Stuff>> b;
-        // Stuff<List<Stuff>> c;
-        // all of the above should be equal
+    protected abstract Set<JavaTypeDefinition> getSuperTypeSet(Set<JavaTypeDefinition> destinationSet);
 
-        for (int i = 0; i < getTypeParameterCount(); ++i) {
-            // Note: we assume that cycles can only exist because of raw types
-            if (!getGenericType(i).equals(otherTypeDef.getGenericType(i))) {
-                return false;
-            }
-        }
+    public abstract Set<Class<?>> getErasedSuperTypeSet();
 
-        return true;
+    /**
+     * @return true if clazz is generic and had not been parameterized
+     */
+    public final boolean isRawType() {
+        return !isIntersectionType() && isGeneric() && CLASS_EXACT_TYPE_DEF_CACHE.containsKey(getType());
```

Comment:
Well it kind of is an optimization, but you are right, we shouldn't rely on it for app logic. Should we use a boolean field instead?